### ISSUE TITLE
Code coverage doesn't work with modules

### ISSUE THREAD

**Issue by sdecker** [pester/Pester#349]
I seem to be able to only get Code Coverage statistics if I dotsource the original scripts. If I have functions that rely on internal functions (not mocked) and I dotsource and use InModuleScope. The tests are excluded from the coverage analysis. I'm guessing this is by design because of how the module scoping works, but I'm hoping I'm just missing something here.

**Comment by dlwyatt** [comment]
It should be working.  Can you share your test script, the folder structure of your module, and the command line you're using when you call Invoke-Pester?

**Comment by sdecker** [comment]
I've tried a couple of variations with the test scripts. Right now our module has a functions subdirectory and a file per function. (This was the only way we could work out how to mock functions with PSUnit which we used to use.) So we have a Tests.ps1 for each function.ps1. We also have them in different trees right. 

+--src
¦  +--DeploymentHelper
¦     +--Functions
¦     ¦  +--AreaManagement
¦     ¦  ¦  +--Get-AreaShortLanguageCode.ps1
¦     ¦  ¦  +--Get-ExpectedODAreaConfiguration_Public.ps1
¦     ¦  ¦  +--Get-ODAreaCfgSetting_Public.ps1
¦     ¦  ¦  +--Get-ODAreasWithQmrFile.ps1
....etc
¦     ¦  +--Common
¦     ¦  ¦  Add-HostFileEntry_Public.ps1
¦     ¦  ¦  Add-32bitODBCDSN_Public.ps1
¦     ¦  ¦  Add-PathVariableAtScope.ps1
¦     ¦  ¦  Add-PathVariable_Public.ps1
¦     ¦  ¦  Get-OSBitness_Public.ps1
....etc

+--test
¦  +--DeploymentDHelper.Test
¦     +--functions
¦     ¦  +--AreaManagement
¦     ¦  ¦  Get-AreaShortLanguageCode.Tests.ps1
¦     ¦  ¦  Get-ExpectedODAreaConfiguration_Public.Tests.ps1
¦     ¦  ¦  Get-ODAreasWithQmrFile.Tests.ps1
¦     ¦  ¦  Invoke-UpdateManifest_Public.Tests.ps1
¦     ¦  ¦  New-UpdateManifest_Public.Tests.ps1
¦     ¦  ¦  Read-UpdateManifest.Tests.ps1
¦     ¦  +--Common
¦     ¦  ¦  Add-PathVariableAtScope.Tests.ps1
¦     ¦  ¦  Add-PathVariable_Public.Tests.ps1
¦     ¦  ¦  Test-OSVersion_Public.Tests.ps1
....etc

If the test scripts call functions that do not use internal functions I can use the following which allows me to produce code coverage stats

$Source = (Split-Path -Parent $PSCommandPath) -replace "\.Test\\","\\" -Replace "test","src"
$sut = (Split-Path -Leaf $PSCommandPath) -Replace ".Tests.", "."
. "$Source\$sut"

Describe "Get-AreaShortLanguageCode" {
}

But if I need to test functions that call internal module functions I don't need to Mock I need to to 

$Source = (Split-Path -Parent $PSCommandPath) -replace "\.Test\\","\\" -Replace "test","src"
$sut = (Split-Path -Leaf $PSCommandPath) -Replace ".Tests.", "."
. "$Source\$sut"

InModuleScope DeploymentHelper {
	$Here = Split-Path -Parent $PSCommandPath

	Describe "Invoke-UpdateManifest" {
}

This file then reports no code coverage on Invoke-UpdateManifest.

I was wonder if we need to merge our .Tests.ps1 files into the same directory structure as the source?

**Comment by dlwyatt** [comment]
You're dot-sourcing the .ps1 files, but using InModuleScope.  That's an odd combination to see, and I suspect that you're running some of your code in a copy of the module that's on your PSModulePath, as a result.  If you're using InModuleScope, I'd expect your Tests scripts to be using Import-Module to load up the code that's to be tested.  If you're dot-sourcing individual ps1 files, there's no need for InModuleScope, as the dot-sourced code will just be out in the main session.

Because you have so many files in this module, I'd recommend importing it once before running Invoke-Pester, and just having the individual Tests.ps1 files assume that the module is already loaded.  You can put this code into something like Invoke-Tests.ps1 out in the root of the test folder, or something like that.  Get rid of all of the `$source` / `$sut` stuff from the tests scripts, as well.

```posh
Remove-Module [D]eploymentHelper
Import-Module $PSScriptRoot\..\src\DeploymentHelper -ErrorAction Stop

$files = Get-ChildItem -Path $PSScriptRoot\..\src -File -Recurse -Include *.ps1,*.psm1

$result = Invoke-Pester -Script $PSScriptRoot -CodeCoverage $files -PassThru

# Further code using $result, if desired.
```

With that sort of code launching Invoke-Pester, as long as you use InModuleScope DeploymentHelper at the necessary points of your test scripts, you should be in business.

**Comment by sdecker** [comment]
Unfortunately, I'm still not having any luck. As an example I have a non-exported function Get-AreaShortLanguageCode.ps1 in the module. 

function Get-AreaShortLanguageCode
{
	[CmdletBinding()]
	param(
		[Parameter(Position=0,Mandatory=$true)]
		[ValidateNotNullOrEmpty()]
		[ValidatePattern('^[\w]+$')]
		[string]$AreaName
	)

    # Get the row from the DB
    try
    {
        $DBDetails = Get-DatabaseDetails Configuration
        $query = ("SELECT CultureCode FROM [Language] INNER JOIN Area ON [Language].LanguageID = Area.LanguageID WHERe Area.AreaName = '{0}'" -f $AreaName)
        $result = Invoke-Sqlcmd -Query $query -ServerInstance $DBDetails.InstanceName -Database $DBDetails.DatabaseName -Username $DBDetails.UserName -Password $DBDetails.Password -OutputSqlErrors $true -Verbose
    }
    catch [System.Exception] 
    {
        $exMsg = ("Exception while getting language short code for area '{0}': {1}" -f $AreaName, $_.Exception.ToString() )
        throw $exMsg
    }

    # Process the result
    if ($result -ne $null)
    {
        $longCode = $result.CultureCode
        $shortCode = $longCode.Split("-")[0]
        return $shortCode
    }
    else 
    {
        throw "Unable to get language short code for area '$AreaName'"
    }
}

I have a test Get-AreaShortLanguageCode.Tests.ps1

#$Here = Split-Path -Parent $PSCommandPath
#$Source = (Split-Path -Parent $PSCommandPath) #-replace "\.Test\\","\\" -Replace "test","src"
#$sut = (Split-Path -Leaf $PSCommandPath) -Replace ".Tests.", "."
#. "$Source\$sut"
Describe "Get-AreaShortLanguageCode" {

  InModuleScope DeploymentHelper {

    It "Should return 'te' for 'te-st'" 	{
      Mock -CommandName Invoke-Sqlcmd {
        $retObj = New-Object PSObject
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "CultureCode" -Value "te-st"
        return $retObj
      }

      Mock -CommandName Get-DatabaseDetails {
        $retObj = New-Object PSObject
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "InstanceName" -Value "testDbServer"
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "DatabaseName" -Value "testDbName"
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "UserName" -Value "testDbUser"
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "Password" -Value "testDbPassword"
        return $retObj
      }

      Get-AreaShortLanguageCode -AreaName "TestArea" | Should Be "te" 
    }

    It "Should throw if no matching area or language code" 	{
      Mock -CommandName Invoke-Sqlcmd { }

      Mock -CommandName Get-DatabaseDetails {
        $retObj = New-Object PSObject
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "InstanceName" -Value "testDbServer"
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "DatabaseName" -Value "testDbName"
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "UserName" -Value "testDbUser"
        Add-Member -InputObject $retObj -MemberType NoteProperty -Name "Password" -Value "testDbPassword"
        return $retObj
      }

      {Get-AreaShortLanguageCode -AreaName "TestArea" }| Should Throw "Unable to get language short code for area 'TestArea'"
    }

    It "Should throw with details if real exception happens" {

      Mock -CommandName Get-DatabaseDetails { Throw "ItBroke" }

      {Get-AreaShortLanguageCode -AreaName "TestArea" }| Should Throw "Exception while getting language short code for area 'TestArea': System.Management.Automation.RuntimeException: ItBroke"
    }


  }
}

I'm trying to test it with the commands

push-location "C:\Temp\UnitTest\solutions\src"
Remove-Module DeploymentHelper
Import-Module .\DeploymentHelper -ErrorAction Stop
$files = Get-ChildItem -Path .\DeploymentHelper -File -Recurse -Include *.ps1,*.psm1 -exclude *.Tests.ps1
$result = Invoke-Pester -CodeCoverage $files -PassThru

The output is 

Describing Get-AreaShortLanguageCode
 [+] Should return 'te' for 'te-st' 231ms
 [+] Should throw if no matching area or language code 45ms
 [+] Should throw with details if real exception happens 31ms
Tests completed in 308ms
Passed: 3 Failed: 0 Skipped: 0 Pending: 0

Code coverage report:
Covered 0.00 % of 28 analyzed commands in 2 files.

Missed commands:

File                                                   Function                  Line Command
----                                                   --------                  ---- -------
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   15 $DBDetails = Get-DatabaseDetai...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   16 $query = ("SELECT CultureCode ...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   16 "SELECT CultureCode FROM [Lang...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   17 $result = Invoke-Sqlcmd -Query...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   21 $exMsg = ("Exception while get...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   21 "Exception while getting langu...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   22 throw $exMsg
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   26 if ($result -ne $null)...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   28 $longCode = $result.CultureCode
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   29 $shortCode = $longCode.Split("...
Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   30 return $shortCode

Functions\AreaManagement\Get-AreaShortLanguageCode.ps1 Get-AreaShortLanguageCode   34 throw "Unable to get language ...

If I change the test file to uncomment the lines to dotsource the script and comment out the InModuleScope. It get the coverage analysis I expect.

Describing Get-AreaShortLanguageCode
 [+] Should return 'te' for 'te-st' 352ms
 [+] Should throw if no matching area or language code 39ms
 [+] Should throw with details if real exception happens 20ms
Tests completed in 412ms
Passed: 3 Failed: 0 Skipped: 0 Pending: 0

Code coverage report:
Covered 100.00 % of 12 analyzed commands in 1 file.

**Comment by dlwyatt** [comment]
Is your module open source?  If I can download it and try this myself, I should be able to figure out what the problem is.

**Comment by sdecker** [comment]
Thank you for you patience working through this. I've figured out the issue. It was a flaw in how we load the subfunctions

We were doing
$BasePath = (Get-Module 'QMODHelper' -ListAvailable).ModuleBase

Which was loading the files from the installed version and not my test version even when I imported the test version as I have the module installed on my local system. 

I changed that line to use 
Split-Path -Path $MyInvocation.MyCommand.Path
And the code coverage works as expected.

Again thank you for taking the time to help me figure this out. I tried all kinds of variations to get it to work and in the end was loading unexpected files.
